api/equality/equality.go

```diff
@@ -27,3 +30,50 @@ func TaskStatusesEqualStable(a, b *api.TaskStatus) bool {
 	copyA.Timestamp, copyB.Timestamp = nil, nil
 	return reflect.DeepEqual(&copyA, &copyB)
 }
+
+// RootCAEqualStable compares RootCAs, excluding join tokens, which are randomly generated
+func RootCAEqualStable(a, b *api.RootCA) bool {
+	switch {
+	case a == nil && b == nil:
+		return true
+	case a != nil && b != nil:
+	default:
+		return false
+	}
+	if !(bytes.Equal(a.CACert, b.CACert) &&
+		subtle.ConstantTimeCompare(a.CAKey, b.CAKey) == 1 &&
+		a.CACertHash == b.CACertHash) {
+		return false
+	}
+	switch {
+	case a.RootRotation == nil && b.RootRotation == nil:
+		return true
+	case a.RootRotation != nil && b.RootRotation != nil:
+	default:
+		return false
+	}
+	return (bytes.Equal(a.RootRotation.CACert, b.RootRotation.CACert) &&
+		subtle.ConstantTimeCompare(a.RootRotation.CAKey, b.RootRotation.CAKey) == 1 &&
+		bytes.Equal(a.RootRotation.CrossSignedCACert, b.RootRotation.CrossSignedCACert))
+}
+
+// ExternalCAsEqualStable compares lists of external CAs and determines whether they are equal
+func ExternalCAsEqualStable(a, b []*api.ExternalCA) bool {
```

Comment:
Yeah, but can this situation ever come up? I believe we are always assigning `s.externalCA` from the protobuf object, so I don't see how an empty map could change to a nil or vice versa.

Is the concern about the first call, before we've initialized anything in the `Server` struct, where `nil` may actually be equivalent to what we would update it to (if the protobuf object contains an empty list)?

In this case, could the function be shortened to:

```go
if len(a) == 0 && len(b) == 0 {
        return true
}
return reflect.DeepEqual(a, b)
```